OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+                       dF3_dprojFibLenNorm) * cosPenn - (activ*F1*F2 + F3) *
+                        dcosPenn_dprojFibLen;
+        double dfRes_dprojFibVelNorm =
+                - activ*F1*dF2_dprojFibVelNorm - dF5_dprojFibVelNorm;
+        double dfRes_dmuscleLength = dF4_dmuscleLength;
+
+
+        //y=(projFiberLength,activation)  <-States
+
+        //df_fy is 2x2:  Where columns are states:
+        //          [projFiberLengthNorm, activation]
+        //              Rows are:
+        //          [force residual; Activation Residual]
+
+        // Row 1 - df_dy  (force)
+        df_dy[0][0] = dfRes_dprojFibLengthNorm;
+        df_dy[0][1] = dfRes_dactiv;
+        // Row 2 - df_dy (activation)
+        df_dy[1][0] = 0;
+        df_dy[1][1] = dActRes_dactiv ;
+
+
+        // Row 1 - df_dydot  (force)
+        df_dydot[0][0] = dfRes_dprojFibVelNorm;
+        df_dydot[0][1] = 0;
+        // Row 2 - df_dydot (activation)
+        df_dydot[1][0] = 0;
+        df_dydot[1][1] = dActRes_dactivdot;
+
+        df_dmuscleLength = dfRes_dmuscleLength;
+
+    }
+
+
+
+    VandenBogert2011Muscle::ImplicitResidual Results;
+
+    Results.forceResidual=fRes;
+    Results.activResidual=activationResidual;
+    Results.forceTendon=Fsee;
+    Results.df_dy=df_dy;
+    Results.df_dydot=df_dydot;
+    Results.df_du=df_du;
+    Results.df_dmuscleLength=df_dmuscleLength;
+    Results.F1=F1;    //Output force components for troubleshooting
+    Results.F2=F2;
+    Results.F3=F3;
+    Results.F4=F4;
+    Results.F5=F5;
+
+
+
+return Results; }
+
+
+
+
+
+
+//------------------------------------------------------------------------------
+VandenBogert2011Muscle::ImplicitResidual
+VandenBogert2011Muscle::calcJacobianByFiniteDiff(double muscleLength,
+                                                 double projFibLenNorm,
+                                                 double activ,
+                                                 double projFibVelNorm,
+                                                 double activdot, double u,
+                                                 double h) const
+{
+    SimTK::Vec2 y;
+    SimTK::Vec2 ydot;
+    y[0]=projFibLenNorm;
+    y[1]=activ;
+    ydot[0]=projFibVelNorm;
+    ydot[1]=activdot;
+
+    // Overload method for state vectors as parameters
+    VandenBogert2011Muscle::ImplicitResidual Results =
+            calcJacobianByFiniteDiff(y, ydot, muscleLength, u, h );
+
+    return Results;
+}
+
+
+//------------------------------------------------------------------------------
+VandenBogert2011Muscle::ImplicitResidual    VandenBogert2011Muscle::
+calcJacobianByFiniteDiff(SimTK::Vec2 y, SimTK::Vec2 ydot, double muscleLength,
+                         double u, double h ) const {
+
+
+
+    //TODO: In the optimizer class, you do not have to supply the Jacobian
+    // function.  The must mean there is code in Simbody that does finite diff.
+    // Should probably look at calling that in place of this code.
```

Comment:
Good idea; you are correct. https://simtk.org/api_docs/simbody/3.5/classSimTK_1_1Differentiator.html